### PR TITLE
[Parse]: Separate do and while blocks diagnostics

### PR DESCRIPTION
**Pull request by kitaisreal**
SR-11148: Separate do and while blocks generate error from legacy diagnostic

1. Taking new line into account if there is a `while` token after stmt.
2. Updated diagnostic to provide new line separator.

More information in ticket https://bugs.swift.org/browse/SR-11148.

### CODE REVIEW

**Comment by CodaFi** [comment]
Could you please add tests?

File: lib/Parse/ParseStmt.cpp [hunk 1]
```diff
@@ -1917,18 +1917,23 @@ ParserResult<Stmt> Parser::parseStmtDo(LabeledStmtInfo labelInfo) {
...
+  SourceLoc whileLoc = Tok.getLoc();
+  if (Tok.isAtStartOfLine()) {
+    return makeParserResult(status,
+                          new (Context) DoStmt(labelInfo, doLoc, body.get()));
+  }
+    // But if we do, advise the programmer that it's 'repeat' now.
   diagnose(doLoc, diag::do_while_now_repeat_while)
-    .fixItReplace(doLoc, "repeat");
+    .fixItReplace(doLoc, "repeat")
+    .fixItInsert(whileLoc, "\n");
```
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp]: I don't think this is right.
If the user applies these fix-its, the code will be something like:
```swift
repeat {
}
while condition
```
The newline is not needed at all.

If we want users to choose between `repeat` and bare `do`, you need to emit 3 diagnostics:
```
error: 'do-while' statement is not allowed.
note: did you mean 'repeat-file' statement?  (with fix-it to 'repeat')
note: did you mean separate 'do' and 'while' statements? (with fix-it inserting '\n')
```

File: lib/Parse/ParseStmt.cpp [hunk 2]
```diff
@@ -1917,18 +1917,23 @@ ParserResult<Stmt> Parser::parseStmtDo(LabeledStmtInfo labelInfo) {
...
+  // from new line. This is just the bare `do` scoping statement.
+  if (Tok.getKind() != tok::kw_while) {
     return makeParserResult(status,
-                         new (Context) DoStmt(labelInfo, doLoc, body.get()));
+                            new (Context) DoStmt(labelInfo, doLoc, body.get()));
   }
-
-  // But if we do, advise the programmer that it's 'repeat' now.
+  SourceLoc whileLoc = Tok.getLoc();
+  if (Tok.isAtStartOfLine()) {
```
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp]: Could you cosolidate these two `if` into one?
```cpp
if (Tok.getKind() != tok::kw_while || Tok.isStartOfLine()) {
    return makeParserResult(...);
}

**Review by rintaro — CHANGES_REQUESTED** [review]
Thank you for working on this! Just a couple of requests:

**Pull request closed without merging**